libraries/chain/block_header_state.cpp

```diff
@@ -141,6 +141,65 @@ const finalizer_policy& block_header_state::get_last_pending_finalizer_policy()
    return *active_finalizer_policy;
 }
 
+// Only defined for core.latest_qc_claim().block_num <= ref.block_num() <= core.current_block_num()
+// Retrieves the finalizer policies applicatble for the block referenced by `ref`.
+// See full explanation in issue #694.
+// ------------------------------------------------------------------------------------------------
+finalizer_policies_t block_header_state::get_finalizer_policies(const block_ref& ref) const {
+   assert(core.links.empty() ||   // called from a bogus block_state constructed in a test
+          (core.latest_qc_claim().block_num <= ref.block_num() && ref.block_num() <= core.current_block_num()));
+   finalizer_policies_t res;
+
+   res.finality_digest = ref.finality_digest;
+
+   auto active_gen = ref.active_policy_generation;
+   assert(active_gen != 0);                                       // we should always have an active policy
+
+   if (active_finalizer_policy->generation == active_gen)
+      res.active_finalizer_policy = active_finalizer_policy;      // the one active at block_num is still active
+   else {
+      // cannot be the pending one as it never was active
+      assert(!pending_finalizer_policy || pending_finalizer_policy->second->generation > active_gen);
+
+      // has to be the one in latest_qc_claim_block_active_finalizer_policy
+      assert(latest_qc_claim_block_active_finalizer_policy != nullptr);
+      assert(latest_qc_claim_block_active_finalizer_policy->generation == active_gen);
+      EOS_ASSERT(latest_qc_claim_block_active_finalizer_policy->generation == active_gen, chain_exception,
```

Comment:
Should be consistent with the asserts above.

This should never happen, but if we somehow are wrong and it does, we would like to avoid the undefined behavior. 

So the condition for the `EOS_ASSERT` should be: `latest_qc_claim_block_active_finalizer_policy != nullptr && latest_qc_claim_block_active_finalizer_policy->generation == active_gen`.